algorithm.go

```diff
@@ -40,31 +40,68 @@ const (
 	Canonical = SHA256
 )
 
+// CryptoHash is the interface that any hash algorithm must implement
+type CryptoHash interface {
+	// Available reports whether the given hash function is usable in the current binary.
+	Available() bool
+	// Size returns the length, in bytes, of a digest resulting from the given hash function.
+	Size() int
+	// New returns a new hash.Hash calculating the given hash function. If the hash function is not
+	// available, it may panic.
+	New() hash.Hash
+}
+
 var (
 	// TODO(stevvooe): Follow the pattern of the standard crypto package for
 	// registration of digests. Effectively, we are a registerable set and
 	// common symbol access.
 
 	// algorithms maps values to hash.Hash implementations. Other algorithms
 	// may be available but they cannot be calculated by the digest package.
-	algorithms = map[Algorithm]crypto.Hash{
-		SHA256: crypto.SHA256,
-		SHA384: crypto.SHA384,
-		SHA512: crypto.SHA512,
-	}
+	algorithms = map[Algorithm]CryptoHash{}
 
 	// anchoredEncodedRegexps contains anchored regular expressions for hex-encoded digests.
 	// Note that /A-F/ disallowed.
-	anchoredEncodedRegexps = map[Algorithm]*regexp.Regexp{
-		SHA256: regexp.MustCompile(`^[a-f0-9]{64}$`),
-		SHA384: regexp.MustCompile(`^[a-f0-9]{96}$`),
-		SHA512: regexp.MustCompile(`^[a-f0-9]{128}$`),
-	}
+	anchoredEncodedRegexps = map[Algorithm]*regexp.Regexp{}
+
+	// algorithmsLock protects algorithms, and anchoredEncodedRegexps
+	algorithmsLock sync.RWMutex
 )
 
+// RegisterAlgorithm may be called to dynamically register an algorithm. The implementation is a CryptoHash, and
+// the regex is meant to match the hash portion of the algorithm. If the algorithm uses a standard hexdigest, the
+// HexDigestRegex helper may be used or the anchorRegexp may be null, and the regex will be derived based on
+// HexDigestRegex. If a duplicate algorithm is already registered, the return value is false, otherwise if registration
+// was successful the return value is true.
+func RegisterAlgorithm(algorithm Algorithm, implementation CryptoHash, anchorRegexp *regexp.Regexp) bool {
+	algorithmsLock.Lock()
+	defer algorithmsLock.Unlock()
+
+	if _, ok := algorithms[algorithm]; ok {
```

Comment:
We made the standard keys lowercase. Perhaps we should force the same here of all keys (and comment as such) so that "SHA256" doesn't confuse with "sha256"

Reply:
@vbatts Should I add a validator based on the semantic here? https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests

What do you think about adding underscores to the BNF to allow for XOFs?